File: Linked-List-Cycle-II/Linked-List-Cycle-II.py

‘’’
 Definition for singly-linked list.
 class ListNode:
     def __init__(self, x):
         self.val = x
         self.next = None
‘’’
class Solution:
    # @param head, a ListNode
    # @return a list node
    def detectCycle(self, head):
        if head==None:
            return None
        if head.next==None:
            return None
        if head.next.next==None:
            return None
        p=head
        q=head.next.next
        while p!=None and q!=None:
            if q.next==None:
                return None
            if q.next.next==None:
                return None
            if q.val==p.val:
                p=head
                q=q.next.next
                while p!=q:
                    p=p.next
                    q=q.next
                return p
            p=p.next
            q=q.next.next
        return None
 

File: solutions/Linked-List-Cycle-II/Linked-List-Cycle-II.cpp

// /**
//  * Definition for singly-linked list.
//  * struct ListNode {
//  *     int val;
//  *     ListNode *next;
//  *     ListNode(int x) : val(x), next(NULL) {}
//  * };
//  */
// class Solution {
// public:
//     ListNode *detectCycle(ListNode *head) {
//         if(!head || !head->next)
//             return NULL;
//         ListNode *slow=head;
//         ListNode *fast=head;
//         while(slow)
//         {
//             slow=slow->next;
//             if(!fast)
//                 return NULL;
//             fast=fast->next;
//             if(!fast)
//                 return NULL;
//             fast=fast->next;
//             if(slow==fast)
//                 break;
//         }
//         if(!slow)
//             return NULL;
//         while(head!=slow)
//         {
//             head=head->next;
//             slow=slow->next;
//         }
//         return head;
//     }
// };

// 2019-07-27
// O(n)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* fast = head;
        ListNode* slow = head;
        while(fast)
        {
            fast = fast->next;
            if(fast)
            {
                fast = fast->next;
            }
            else
            {
                return NULL;
            }
            slow = slow->next;
            if(fast == slow)
            {
                while(head != slow)
                {
                    head = head->next;
                    slow = slow->next;
                }
                return slow;
            }
        }
        return NULL;
    }
};
 